File: server/app/models/item.py
from sqlalchemy.orm import relationship

from app.database import Base

from sqlalchemy import Column, Integer, String, TIMESTAMP, func, Float, ForeignKey


class Item(Base):
    __tablename__ = "items"

    id = Column(Integer, primary_key=True, index=True)
    slug = Column(String, index=True)
    name = Column(String, nullable=False)
    description = Column(String, nullable=True)

    # category map to category table
    category_id = Column(Integer, ForeignKey('categories.id'), nullable=False)
    category = relationship("Category", foreign_keys=[category_id], lazy="joined")

    # This is a JSON string
    images = Column(String, nullable=True, default="[]")
    thumbnail = Column(String, nullable=True)

    # item stat
    historical_sold = Column(Integer, default=0)
    rating_count = Column(Integer, default=0)
    rating_avg = Column(Float, default=0)
    comment_count = Column(Integer, default=0)
    price_before_discount = Column(Float, default=0)

    provider = Column(String, nullable=False)
    provider_url = Column(String, nullable=False)

    created_at = Column(TIMESTAMP(timezone=True),
                        nullable=False, server_default=func.now())
    updated_at = Column(TIMESTAMP(timezone=True),
                        default=None, onupdate=func.now())

File: client/types/item.ts
import type { BlockRsp } from '~/types/base';

export type ItemDetailRsp = BlockRsp<ItemDetail>;

export type ItemDetail = {
    id: number;
    slug: string;
    description: string;
    images: string[];
    historical_sold: number;
    rating_avg: number;
    price_before_discount: number;
    provider_url: string;
    updated_at: null;
    name: string;
    category_id: number;
    thumbnail: string;
    rating_count: number;
    comment_count: number;
    provider: string;
    created_at: Date;
    category: Category;
    categories: Category[];
    current_price: number;
};

export type Category = {
    level: number;
    id: number;
    created_at: Date;
    name: string;
    slug: string;
    parent_id: number | null;
    updated_at: Date | null;
};

export type PriceHistory = {
    price: number;
    id: number;
    item_id: number;
    created_at: Date;
};

export type PriceHistoryRsp = BlockRsp<PriceHistory[]>;
